src/chains/filecoin/filecoin/src/blockchain.ts

```diff
@@ -213,7 +218,43 @@ export default class Blockchain extends Emittery.Typed<
       timeout: 500 // Enforce a timeout; otherwise will hang if CID not found
     });
 
-    return stat.DataSize;
+    return stat.CumulativeSize;
+  }
+
+  private async downloadFile(cid: string, ref: FileRef): Promise<void> {
+    if (!this.ipfsServer.node) {
+      throw new Error("IPFS server is not running");
+    }
+
+    const dirname = path.dirname(ref.path);
+    try {
+      if (!fs.existsSync(dirname)) {
+        await fs.promises.mkdir(dirname, { recursive: true });
+      }
+      await fs.promises.writeFile(ref.path, "");
+    } catch (e) {
+      throw new Error(
+        `Could not create file.\n  CID: ${cid}\n  Path: ${
+          ref.path
+        }\n  Error: ${e.toString()}`
+      );
+    }
+
+    const chunks = this.ipfsServer.node.files.read(new IPFS_CID(cid), {
+      timeout: 500 // Enforce a timeout; otherwise will hang if CID not found
+    });
+
+    for await (const chunk of chunks) {
+      try {
+        await fs.promises.appendFile(ref.path, chunk, "binary");
```

Comment:
1. Wouldn't this just cause a data file corruption, which is hypothetically implied if you stop a service in the middle of saving data? I feel like we don't care about this scenario. The user may be equally confused if they stop the service and the file isn't there at all. If we really wanted to prevent confusion, the proper way is to scope creep (because it hasn't been used in any of the Filecoin dapps I've demoed and not in the support plan) and implement the subscription method `Filecoin.ClientRetrieveWithEvents` which dapps/users can monitor the progress of the retrieval and know whether or not the file was completely retrieved before shutting it down (or see that it was only partially retrieved when they can concur the file only partially exists).
2. It could help on performance for very large files. It could also open up a new issue related to partial downloads as what happens if the stream doesn't properly close. We can make this change if you want it; I personally don't think it's necessary at this time.
3. I consider the directory creation a `no-op` in terms of performance as it doesn't really affect Big O. I think the code is more readable from a logical stand point as is: first make sure the directory exists, then touch the file, then fill the file in chunks. However we can make this change if you really want it.

Reply:
> If we do care, should we instead write to a tmp file first then move that the ile to ref.path 😬 (there are additional gotchas if we tried this, as moving files across a partition is slow and tricky)?

After thinking more, I believe this just reintroduces the problem. Moving files in some cases (i.e. across partiations) is the same as copying and deleting, rather than just changing the inode location, in which you would basically be doubling the file transfer ops. If we wanted to do this, we could follow the pattern browsers use during download and put the file in the same location, but give it a suffix or prefix saying it's a partial file and rename it when it's complete

Reply:
1. It wouldn't cause corruption for many file types, the files would just be missing some information at the end (text files, some databases, some video and image formats, encrypted files, etc). An example of a popular implementation (not saying it's exactly right for us): Chrome names the file `{filename}.{ext}.crdownload` until it is done downloading, but that's likely just used for chrome's download manager pause/resume stuff. I'm okay just keeping the partially downloaded file if you are.
2. Not sure what you mean by `the stream doesn't partially close`. We just need to make sure we close it, no? Windows antivirus locking the file is still a concern I have here.
3. I agree on your point about readability, but also know that teeecchhnniiically it's not "correct", since it is still possible for the directory (and that empty file) to be deleted before we open a handle to the file. Oh, and I think `appendFile` creates the file (but not the directory) already.

Reply:
I went ahead and implemented the `.partial` suffix as well as the WriteStream. I believe these changes should resolve each of your concerns to some degree. Let me know if it's not enough!